File: api/src/main/java/com/hirrao/javaee/interceptors/LoginInterceptor.java
package com.hirrao.javaee.interceptors;

import com.hirrao.javaee.entity.User;
import com.hirrao.javaee.service.UserService;
import com.hirrao.javaee.utils.Jwt;
import com.hirrao.javaee.utils.ThreadLocalUtil;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.HandlerInterceptor;


//拦截器
@Component
public class LoginInterceptor implements HandlerInterceptor {
    private final UserService userService;
    private static final Logger logger = LoggerFactory.getLogger(LoginInterceptor.class);

    @Autowired
    private LoginInterceptor(UserService userService) {
        this.userService = userService;
    }

    @Override
    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
        if ("OPTIONS".equalsIgnoreCase(request.getMethod())) {
            return true;
        }
        logger.debug(request.getHeader("Authorization"));
        //获取请求头中的token
        String token = request.getHeader("Authorization");
        logger.debug("token={}", token);
        //先验证token
        try {
            String Uid = Jwt.parseToken(token);
            logger.debug(Uid);
            User user = userService.findByUid(Long.parseLong(Uid));
            ThreadLocalUtil.set(user);
            return true;
        } catch (Exception e) {
            //响应状态码为401
            response.setStatus(401);
            return false;
        }
    }

    @Override
    public void afterCompletion(HttpServletRequest request, HttpServletResponse response, Object handler, Exception ex) {
        //清除ThreadLocal中的用户信息
        ThreadLocalUtil.remove();
    }
}

File: api/src/main/kotlin/com/hirrao/javaee/interceptors/LoginInterceptor.kt
package com.hirrao.javaee.interceptors

import com.hirrao.javaee.service.UserService
import com.hirrao.javaee.utils.Jwt.parseToken
import com.hirrao.javaee.utils.ThreadLocalUtil.remove
import com.hirrao.javaee.utils.ThreadLocalUtil.set
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.lang.Nullable
import org.springframework.stereotype.Component
import org.springframework.web.servlet.HandlerInterceptor

//拦截器
@Component
class LoginInterceptor @Autowired constructor(private val userService: UserService) : HandlerInterceptor {
    private val logger: Logger = LoggerFactory.getLogger(javaClass)

    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
        if ("OPTIONS".equals(request.method, ignoreCase = true)) {
            return true
        }
        println(request.getHeader("Authorization"))
        //获取请求头中的token
        val token = request.getHeader("Authorization")
        logger.debug("token={}", token)
        //先验证token
        try {
            val uid = parseToken(token)
            println(uid)
            val user = userService.findByUid(uid.toInt())
            set(user!!)
            return true
        } catch (e: Exception) {
            //响应状态码为401
            response.status = 401
            return false
        }
    }

    override fun afterCompletion(
        request: HttpServletRequest, response: HttpServletResponse, handler: Any, @Nullable ex: java.lang.Exception?
    ) {
        //清除ThreadLocal中的用户信息
        remove()
    }
}